classes/local/rule_instances_controller.php

```diff
@@ -168,18 +168,72 @@ public function get_rule_instance_records(): array {
     }
 
     /**
-     * Return the rule instance record matching the given instanceid.
+     * Return a list of hydrated rule instance objects.
+     *
+     * @return array
+     */
+    public function get_rule_instances(): array {
+        $instances = [];
+        foreach ($this->get_rule_instance_records() as $instance) {
+            $pluginrule = 'registrationrule_' . $instance->type . '\rule';
+
+            // Parse additional config and add to instance.
+            foreach (json_decode($instance->other) as $configkey => $configvalue) {
+                $instance->$configkey = $configvalue;
+            }
+            $ruleinstance = new $pluginrule($instance);
+            if (!$ruleinstance instanceof rule\rule_base) {
+                debugging("Rule $pluginrule does not extend rule_base", DEBUG_DEVELOPER);
+                continue;
+            }
+            $instances[$ruleinstance->get_id()] = $ruleinstance;
+        }
+
+        return $instances;
+    }
```

Comment:
The counter argument is whether we need a `rule_interface` at all? Are we more interested in whether it is _actually_ a duck, or if it just quacks like a duck? :)

I'm not too bothered either way, but currently we have an interface that doesn't define all the methods of `rule_base`, so if we just check for the interface then we probably need to improve the interface to include all expected properties.

Reply:
Big plus for the  interface would be: it's clearly defined what we expect and in which way exactly. If we just query for the methods, we just know if we can call them at all. But we don't know if they consume our expected parameters and provide expected return values (and types).

Reply:
I've implemented this and I will push as a new commit in the morning for you to see, but it has left me confused about whether this is a good idea.

The PHP docs state...


> Although they are supported, including [constructors](https://www.php.net/manual/en/language.oop5.decon.php#language.oop5.decon.constructor) in interfaces is strongly discouraged. Doing so significantly reduces the flexibility of the object implementing the interface. Additionally, constructors are not enforced by inheritance rules, which can cause inconsistent and unexpected behavior.

We kind of enforce that plugin classes are instantiated with particular parameters and the only way for a developer to know this would be to extend from our abstract base class, unless we include the constructor in the interface, but this is strongly discouraged and wouldn't be enforced by PHP's inheritance rules anyway.

We could just not include the constructor in the interface, but I can't see why anyone would _not_ want to use our abstract class either.

Reply:
In other words, the constructor is an implementation detail and shouldn't be part of an interface, if we need to enforce an implementation detail then we should at least use an abstract class.

I'm actually thinking we should probably enforce the use of our abstract class and maybe even do away with the rule_interface entirely as it seems superfluous. The instances controller needs a rule plugin class to have certain implementation details (i.e. the constructor signature), not just present certain methods. 

We could do this but also create separate interfaces for pre/post data functions and extending the sign-up form. That way we have a proper (not abstract) base class and a range of optional interfaces that rule plugin classes could choose to implement?

Of course, there are conflicting opinions online about this so I'm unable to form a consensus. I'm interested in hearing your thoughts 😁

Reply:
As discussed above I've gone ahead and removed the `rule_interface`, moving abstract rule_base methods to optional interfaces.

The result is in a new commit... `Move abstract rule_base methods to optional interfaces`